packages/warning/src/babel-plugin.js

```diff
@@ -0,0 +1,62 @@
+const pkg = require( '../package.json' );
+
+function babelPlugin( { types: t } ) {
+	const seen = Symbol();
+
+	const binaryExpression = t.binaryExpression(
+		'!==',
+		t.memberExpression(
+			t.memberExpression( t.identifier( 'process' ), t.identifier( 'env' ), false ),
+			t.identifier( 'NODE_ENV' ),
+			false
+		),
+		t.stringLiteral( 'production' )
+	);
+
+	return {
+		visitor: {
+			ImportDeclaration( path, state ) {
+				const { node } = path;
+				const isThisPackageImport = node.source.value.indexOf( pkg.name ) !== -1;
+
+				if ( ! isThisPackageImport ) {
+					return;
+				}
+
+				const defaultSpecifier = node.specifiers.find(
+					( specifier ) => specifier.type === 'ImportDefaultSpecifier'
+				);
+
+				const { name } = defaultSpecifier.local;
+
+				state.callee = name;
+			},
+			CallExpression( path, state ) {
+				const { node } = path;
+
+				// Ignore if it's already been processed
+				if ( node[ seen ] ) {
+					return;
+				}
+
+				const name = state.callee || state.opts.callee;
+
+				if ( path.get( 'callee' ).isIdentifier( { name } ) ) {
+					// Turns this code:
+					// warning(condition, argument, argument);
+					// into this:
+					// process.env.NODE_ENV !== "production" ? warning(condition, argument, argument) : void 0;
```

Comment:
I'm very glad you are thinking about dead code elimination with this pull request.

One concern I have about this specific implementation is that it makes a few assumptions about the runtime and build environments that I'm not sure we should be comfortable to make:

1. Something else in the build will ensure that `process.env.NODE_ENV` global is assigned, or that the generated code is only able to run in Node.js (and not in a browser, where `process.env` does not exist)
2. Something else in the build will perform dead code elimination on unreachable code (like Webpack + Uglify/Terser).

If either of these are not met, then either the code will not in-fact be removed, or at worst it can cause errors in the runtime environment (e.g. trying to access property `NODE_ENV` of an undefined `process.env` in a browser).

What I'd wonder is: Is there any reason we'd not want to just remove the call expression here and now based on what `process.env` evaluates to in the Babel transform? I guess I can see how we might not want that though if, for example, we want the warnings to be logged in projects which depend on our packages (where the published packages would be built in a "production" environment?).

It all looks very similar to Facebook's internal `invariant` utility. I guess this is intentional? How do they handle it? Or, since those are used in internal projects, are they fine to make those assumptions about the runtime/build environment?

Alternatively, we could do some combination of:

- Document the build environment assumptions
- Check that `process` global exists before evaluating it
  - e.g. `typeof process !== 'undefined' && typeof process.env !== 'undefined' && process.env.NODE_ENV`

Reply:
@aduth I think I assumed we were handling `process.env.NODE_ENV` somehow because it's being used in other parts of the code base. For example: https://github.com/WordPress/gutenberg/blob/8719578ea93cc2c51d843627d5a50910cf6c65d7/packages/components/src/tooltip/index.js#L147-L150

Looking at the code generated in the `build`/`build-module` folders, that check remains there. I'm confused now! How does it work? Where can I find the code responsible for building the code for the browser?

> What I'd wonder is: Is there any reason we'd not want to just remove the call expression here and now based on what process.env evaluates to in the Babel transform? I guess I can see how we might not want that though if, for example, we want the warnings to be logged in projects which depend on our packages (where the published packages would be built in a "production" environment?).

I expect some warnings like this:

```js
warning(
  dialog.tabIndex === undefined || dialog.tabIndex < 0,
  'It\'s recommended to have at least one tabbable element inside dialog. The dialog element has been automatically focused.',
  'If this is the intended behavior, pass `tabIndex={0}` to the dialog element to disable this warning.'
);
```

If we remove the warning call from the code we ship to npm, then people won't see that (when using `@wordpress/components`, for example).

Reply:
Facebook uses `__DEV__`:

https://github.com/facebook/react/blob/3c54df0914f02fed146faa519a5a899d0e3af32e/packages/shared/consoleWithStackDev.js#L31

I guess the benefit is that it fallbacks to `undefined` and therefore doesn't call the guarded code. You can still benefit from dead code elimination by teaching webpack to convert it to `true` or `false`.

I don't know what's the best way to move forward, but I'm sharing how I see the approach FB took.

Reply:
> I guess the benefit is that it fallbacks to `undefined` and therefore doesn't call the guarded code. You can still benefit from dead code elimination by teaching webpack to convert it to `true` or `false`.

In this case, I assume something must still be assigning `__DEV__` value, as otherwise the reference to an undeclared variable would throw an error:

```js
( function() {
    if ( __DEV__ ) {
        console.log( 'Constant assigned' );
    } else {
        console.log( 'Constant not assigned' );
    }
}() )
// Uncaught ReferenceError: __DEV__ is not defined
```

Revisiting my earlier comment, I'm actually not sure we're doing anything wrong here with how we reference `process.env.NODE_ENV`. I think it could be fair to say that if we're distributing this as a Node package, it may be more-or-less assumed that it be evaluated as in a Node context where the `process` global is defined. I do worry that in a standalone context, there is a chance that this might not be substituted correctly when building for a browser target. At least in the case of Webpack, this is handled automatically as far as I can tell (see [`mode`](https://webpack.js.org/configuration/mode/), [`DefinePlugin`](https://webpack.js.org/plugins/define-plugin/)).

A safe option might just be to test its presence before evaluating it (`typeof process !== 'undefined'`).

With regards to other existing code in the codebase, I think one of two of the following might be the case:

- We're okay to assume the code is not used outside of Gutenberg, or at least in a context not already subjected to Gutenberg's own build process
- Or, that code is prone to issues as well

For something like a `warning` package, it seemed highly likely we'd want to promote that this be used separate from Gutenberg and its build, hence the original concern.